src/Compilers/VisualBasic/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.vb

```diff
@@ -0,0 +1,303 @@
+
+            ' Do not instrument the instrumentation helpers if they are part of the current compilation (which occurs only during testing). GetCreatePayload will fail with an infinite recursion if it Is instrumented.
+            If createPayload IsNot Nothing AndAlso Not method.IsImplicitlyDeclared AndAlso Not method.Equals(createPayload) Then
+                Return New DynamicAnalysisInjector(method, methodBody, factory, createPayload, diagnostics, debugDocumentProvider, previous)
+            End If
+
+            Return Nothing
+        End Function
+
+        Public ReadOnly Property DynamicAnalysisSpans As ImmutableArray(Of SourceSpan)
+            Get
+                Return _dynamicAnalysisSpans
+            End Get
+        End Property
+
+        Private Sub New(method As MethodSymbol, methodBody As BoundStatement, factory As SyntheticBoundNodeFactory, createPayload As MethodSymbol, diagnostics As DiagnosticBag, debugDocumentProvider As DebugDocumentProvider, previous As Instrumenter)
+            MyBase.New(previous)
+            _createPayload = createPayload
+            _method = method
+            _methodBody = methodBody
+            _spansBuilder = ArrayBuilder(Of SourceSpan).GetInstance()
+            Dim payloadElementType As TypeSymbol = factory.SpecialType(SpecialType.System_Boolean)
+            _payloadType = ArrayTypeSymbol.CreateVBArray(payloadElementType, ImmutableArray(Of CustomModifier).Empty, 1, factory.Compilation.Assembly)
+            _methodPayload = factory.SynthesizedLocal(_payloadType, kind:=SynthesizedLocalKind.InstrumentationPayload, syntax:=methodBody.Syntax)
+            _diagnostics = diagnostics
+            _debugDocumentProvider = debugDocumentProvider
+            _methodHasExplicitBlock = MethodHasExplicitBlock(method)
+            _factory = factory
+
+            ' The first point indicates entry into the method And has the span of the method definition.
+            _methodEntryInstrumentation = AddAnalysisPoint(MethodDeclarationIfAvailable(methodBody.Syntax), factory)
+        End Sub
+
+        Public Overrides Function CreateBlockPrologue(original As BoundBlock, ByRef synthesizedLocal As LocalSymbol) As BoundStatement
+            Dim previousPrologue As BoundStatement = MyBase.CreateBlockPrologue(original, synthesizedLocal)
+
+            If _methodBody Is original Then
+                _dynamicAnalysisSpans = _spansBuilder.ToImmutableAndFree()
+                ' In the future there will be multiple analysis kinds.
+                Const analysisKind As Integer = 0
+
+                Dim modulePayloadType As ArrayTypeSymbol = ArrayTypeSymbol.CreateVBArray(_payloadType, ImmutableArray(Of CustomModifier).Empty, 1, _factory.Compilation.Assembly)
+
+                ' Synthesize the initialization of the instrumentation payload array, using concurrency-safe code
+                '
+                ' Dim payload = PID.PayloadRootField[methodIndex]
```

Comment:
OK. I wish we had enough of a bilingual factory to enable having only one version of this code.